pkg/controller/iam/managed_test.go

```diff
@@ -0,0 +1,652 @@
+						Namespace: namespace,
+						Name:      providerSecretName,
+					},
+					Key: providerSecretKey,
+				},
+			},
+		},
+	}
+
+	secret := corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{Namespace: namespace, Name: providerSecretName},
+		Data:       map[string][]byte{providerSecretKey: []byte(providerSecretData)},
+	}
+
+	type strange struct {
+		resource.Managed
+	}
+
+	type args struct {
+		ctx context.Context
+		mg  resource.Managed
+	}
+	type want struct {
+		err error
+	}
+
+	cases := map[string]struct {
+		conn managed.ExternalConnecter
+		args args
+		want want
+	}{
+		"Connected": {
+			conn: &connecter{
+				client: &test.MockClient{MockGet: func(_ context.Context, key client.ObjectKey, obj runtime.Object) error {
+					switch key {
+					case client.ObjectKey{Name: providerName}:
+						*obj.(*gcpv1alpha3.Provider) = provider
+					case client.ObjectKey{Namespace: namespace, Name: providerSecretName}:
+						*obj.(*corev1.Secret) = secret
+					}
+					return nil
+				}},
+				newIAM: func(_ context.Context, _ []byte) (ProviderClient, error) { return nil, nil },
+			},
+			args: args{
+				ctx: context.Background(),
+				mg:  serviceAccount(),
+			},
+			want: want{
+				err: nil,
+			},
+		},
+		"NotServiceAccount": {
+			conn: &connecter{},
+			args: args{ctx: context.Background(), mg: &strange{}},
+			want: want{err: errors.New(errNotServiceAccount)},
+		},
+		"FailedToGetProvider": {
+			conn: &connecter{
+				client: &test.MockClient{MockGet: func(_ context.Context, key client.ObjectKey, obj runtime.Object) error {
+					return errorBoom
+				}},
+			},
+			args: args{ctx: context.Background(), mg: serviceAccount()},
+			want: want{err: errors.Wrap(errorBoom, errGetProvider)},
+		},
+		"FailedToGetProviderSecret": {
+			conn: &connecter{
+				client: &test.MockClient{MockGet: func(_ context.Context, key client.ObjectKey, obj runtime.Object) error {
+					switch key {
+					case client.ObjectKey{Name: providerName}:
+						*obj.(*gcpv1alpha3.Provider) = provider
+					case client.ObjectKey{Namespace: namespace, Name: providerSecretName}:
+						return errorBoom
+					}
+					return nil
+				}},
+			},
+			args: args{ctx: context.Background(), mg: serviceAccount()},
+			want: want{err: errors.Wrap(errorBoom, errGetProviderSecret)},
+		},
+		"ProviderSecretNil": {
+			conn: &connecter{
+				client: &test.MockClient{MockGet: func(_ context.Context, key client.ObjectKey, obj runtime.Object) error {
+					switch key {
+					case client.ObjectKey{Name: providerName}:
+						nilSecretProvider := provider
+						nilSecretProvider.SetCredentialsSecretReference(nil)
+						*obj.(*gcpv1alpha3.Provider) = nilSecretProvider
+					case client.ObjectKey{Namespace: namespace, Name: providerSecretName}:
+						return errorBoom
+					}
+					return nil
+				}},
+			},
+			args: args{ctx: context.Background(), mg: serviceAccount()},
+			want: want{err: errors.New(errProviderSecretRef)},
+		},
+		"FailedToCreateCloudMemorystoreClient": {
```

Comment:
This looks like some leftover copy pasta here :)

Reply:
✂️ ty